### ISSUE TITLE
Shutdown after plug set to turn on automatically

### ISSUE THREAD

**Issue by fentonmatt** [bwp91/homebridge-platform-wemo#213]
**What issue do you have? Please be as thorough and explicit as possible.**
I have WeMo plug set to turn on hour before sunrise everyday through HK automation and always has worked well. Today when automation tried to turn it on, this happened after it turned on successfully.
...

**Details of your setup.**
Do you use Homebridge or HOOBS?
HOOBS
...

Which version of Homebridge/HOOBS do you have?
HOOBS version - 3.2.9
Node version - 12.13.1
...

Which version of this plugin (homebridge-platform-wemo) do you have?
2.1.2
...

**Please paste any relevant logs below.**
> If you are posting an error then it is helpful for me to also see the previous few lines as this can show the cause of the error.

“11/1/2020, 7:41:22 AM [Belkin WeMo] [Home Spare Outlet] setting state and outlet-in-use to [on].
11/1/2020, 7:41:22 AM [Belkin WeMo] [Home Spare Outlet] setting state and outlet-in-use to [on].
11/1/2020, 7:41:22 AM Error: This callback function has already been called by someone else; it can only be called one time.
    at /usr/local/lib/node_modules/@hoobs/hoobs/node_modules/hap-nodejs/dist/lib/util/once.js:12:19
    at /home/hoobs/.hoobs/node_modules/homebridge-platform-wemo/lib/device/outlet.js:38:9
    at module.exports.<anonymous> (/home/hoobs/.hoobs/node_modules/homebridge-platform-wemo/lib/wemo-client.js:122:5)
    at Parser.<anonymous> (/home/hoobs/.hoobs/node_modules/xml2js/lib/parser.js:304:18)
    at Parser.emit (events.js:210:5)
    at SAXParser.onclosetag (/home/hoobs/.hoobs/node_modules/xml2js/lib/parser.js:262:26)
    at emit (/home/hoobs/.hoobs/node_modules/sax/lib/sax.js:624:35)
    at emitNode (/home/hoobs/.hoobs/node_modules/sax/lib/sax.js:629:5)
    at closeTag (/home/hoobs/.hoobs/node_modules/sax/lib/sax.js:889:7)
    at SAXParser.write (/home/hoobs/.hoobs/node_modules/sax/lib/sax.js:1436:13)
    at Parser.exports.Parser.Parser.parseString (/home/hoobs/.hoobs/node_modules/xml2js/lib/parser.js:323:31)
    at Parser.parseString (/home/hoobs/.hoobs/node_modules/xml2js/lib/parser.js:5:59)
    at Object.exports.parseString (/home/hoobs/.hoobs/node_modules/xml2js/lib/parser.js:369:19)
    at IncomingMessage.<anonymous> (/home/hoobs/.hoobs/node_modules/homebridge-platform-wemo/lib/wemo-client.js:66:16)
    at IncomingMessage.emit (events.js:215:7)
    at endReadableNT (_stream_readable.js:1184:12)
    at processTicksAndRejections (internal/process/task_queues.js:80:21)
11/1/2020, 7:41:22 AM Got SIGTERM, shutting down Bridge...”
...

**Comment by bwp91** [comment]
Hi @fentonmatt 
This issue has me scratching my head!
* Did *anything* change yesterday, e.g., firmware/hoobs/plugin updates?
* Do you get this error when using Home app/HOOBS to switch on/off the device?
* If you were to set up a test automation to turn it on eg in 10 minutes time (from reading this), do you receive the same error?

Thanks

**Comment by fentonmatt** [comment]
Hey @bwp91 thanks for the reply. As far as my HOOBS system goes, I did absolutely nothing yesterday to it that I know of to cause this. No updates, adjustments or anything done to HOOBS or HK. I cannot even get this problem to recreate itself in the same setting as it was the day it happened. All I see is that for some reason the signal to turn the device on from my automation showed twice in my logs and it usually only shows once "switching state to on" and that's it. The first time seems to have worked and the second time gave that error and caused a shutdown. Somehow looks like it got the signal to turn on twice and that wasn't allowed and caused it to shutdown? Only other thing I can think caused it could have been the time change. Yesterday was DST end. But I did nothing to settings myself and HOOBS did that automatically. Only other minor thing was I did some changes to my home network and gave all my devices static IP addresses instead of automatic and messed with that from my eero but haven't noticed any difference with any of my devices from that. Even today, that same automation that happens once a day for on and once at night for off, both times did not cause an issue. Has only happened once but was so weird I wanted to let you know about it. So far has not happened again and I cannot recreate.

**Comment by fentonmatt** [comment]
@bwp91 Error does not happen when manually turning on/off device and it does not happen again when setting a new automation either. All is working perfect today but nothing has changed from that day either.

**Comment by sanist1** [comment]
Same issue here. Once a day, as a result of automation get the same crash. Turning on wemo device manually has no side effect.
Occurs since upgrading to Platform Wemo 2.1.2

11/3/2020, 5:11:38 PM [BelkinWeMo] [Wemo] setting state and outlet-in-use to [on].
11/3/2020, 5:11:38 PM [BelkinWeMo] [Wemo] setting state and outlet-in-use to [on].
11/3/2020, 5:11:38 PM Error: This callback function has already been called by someone else; it can only be called one time.
    at /usr/local/lib/node_modules/@hoobs/hoobs/node_modules/hap-nodejs/dist/lib/util/once.js:12:19
    at /home/hoobs/.hoobs/node_modules/homebridge-platform-wemo/lib/device/outlet.js:38:9
    at module.exports.<anonymous> (/home/hoobs/.hoobs/node_modules/homebridge-platform-wemo/lib/wemo-client.js:122:5)
    at Parser.<anonymous> (/home/hoobs/.hoobs/node_modules/xml2js/lib/parser.js:304:18)
    at Parser.emit (events.js:310:20)
    at SAXParser.onclosetag (/home/hoobs/.hoobs/node_modules/xml2js/lib/parser.js:262:26)
    at emit (/home/hoobs/.hoobs/node_modules/sax/lib/sax.js:624:35)
    at emitNode (/home/hoobs/.hoobs/node_modules/sax/lib/sax.js:629:5)
    at closeTag (/home/hoobs/.hoobs/node_modules/sax/lib/sax.js:889:7)
    at SAXParser.write (/home/hoobs/.hoobs/node_modules/sax/lib/sax.js:1436:13)
    at Parser.exports.Parser.Parser.parseString (/home/hoobs/.hoobs/node_modules/xml2js/lib/parser.js:323:31)
    at Parser.parseString (/home/hoobs/.hoobs/node_modules/xml2js/lib/parser.js:5:59)
    at Object.exports.parseString (/home/hoobs/.hoobs/node_modules/xml2js/lib/parser.js:369:19)
    at IncomingMessage.<anonymous> (/home/hoobs/.hoobs/node_modules/homebridge-platform-wemo/lib/wemo-client.js:66:16)
    at IncomingMessage.emit (events.js:322:22)
    at endReadableNT (_stream_readable.js:1187:12)
    at processTicksAndRejections (internal/process/task_queues.js:84:21)

**Comment by bwp91** [comment]
Do either of you own any other Wemo devices apart from the outlet?

**Comment by fentonmatt** [comment]
@bwp91 I only own this one WeMo outlet and my problem so far since the initial issue has not happened again. So far so good in my end since then. Still nothing has changed before or since but is working.

**Comment by sanist1** [comment]
This is my only Wemo device. I've set up a crontab to restart the homebridge service to counteract the crash for the moment. Will keep an eye on the logfile to see if it is still crashing, though.

**Comment by fentonmatt** [comment]
Randomly happened again tonight this time at the time I have set to turn it off instead of on. Seems to be completely random. 

11/8/2020, 12:59:59 AM [Belkin WeMo] [Home Spare Outlet] setting state and outlet-in-use to [off].
11/8/2020, 12:59:59 AM [Belkin WeMo] [Home Spare Outlet] setting state and outlet-in-use to [off].
11/8/2020, 12:59:59 AM Error: This callback function has already been called by someone else; it can only be called one time.
    at /usr/local/lib/node_modules/@hoobs/hoobs/node_modules/hap-nodejs/dist/lib/util/once.js:12:19
    at /home/hoobs/.hoobs/node_modules/homebridge-platform-wemo/lib/device/outlet.js:38:9
    at module.exports.<anonymous> (/home/hoobs/.hoobs/node_modules/homebridge-platform-wemo/lib/wemo-client.js:122:5)
    at Parser.<anonymous> (/home/hoobs/.hoobs/node_modules/xml2js/lib/parser.js:304:18)
    at Parser.emit (events.js:210:5)
    at SAXParser.onclosetag (/home/hoobs/.hoobs/node_modules/xml2js/lib/parser.js:262:26)
    at emit (/home/hoobs/.hoobs/node_modules/sax/lib/sax.js:624:35)
    at emitNode (/home/hoobs/.hoobs/node_modules/sax/lib/sax.js:629:5)
    at closeTag (/home/hoobs/.hoobs/node_modules/sax/lib/sax.js:889:7)
    at SAXParser.write (/home/hoobs/.hoobs/node_modules/sax/lib/sax.js:1436:13)
    at Parser.exports.Parser.Parser.parseString (/home/hoobs/.hoobs/node_modules/xml2js/lib/parser.js:323:31)
    at Parser.parseString (/home/hoobs/.hoobs/node_modules/xml2js/lib/parser.js:5:59)
    at Object.exports.parseString (/home/hoobs/.hoobs/node_modules/xml2js/lib/parser.js:369:19)
    at IncomingMessage.<anonymous> (/home/hoobs/.hoobs/node_modules/homebridge-platform-wemo/lib/wemo-client.js:66:16)
    at IncomingMessage.emit (events.js:215:7)
    at endReadableNT (_stream_readable.js:1184:12)
    at processTicksAndRejections (internal/process/task_queues.js:80:21)
11/8/2020, 12:59:59 AM Got SIGTERM, shutting down Bridge...

**Comment by bwp91** [comment]
I have a beta version of the plugin available to test. It might not fix the "double update" but it should catch the error and stop Homebridge from restarting each time.

Problem is I don't know the best way for someone to update to a beta version on HOOBS.

**Comment by fentonmatt** [comment]
Yeah I just saw you have a beta version available. Is that not able to be installed via terminal somehow? If not it isn't a major issue. Just a minor nuisance. I can wait for official release and keep you posted. Am down to test if needed/possible on HOOBS as well.

**Comment by bwp91** [comment]
I've updated my "beta version" wiki page with instructions for HOOBS users, if you want to have a try with the beta:

https://github.com/bwp91/homebridge-platform-wemo/wiki/Beta-Version

**Comment by sanist1** [comment]
Hi thanks for the updates. I just installed the Beta version on HOOBS and disabled the messy cron job. I'll let you know if it solves the crash as soon as the automation runs.

**Comment by sanist1** [comment]
hi there. Update on my last: all worked flawlessly since the beta installation. thanks

**Comment by bwp91** [comment]
Oh good news. Anything interesting in the logs at the time of the automation?

**Comment by sanist1** [comment]
... as boring as could be: [BelkinWeMo] [Wemo] setting state and outlet-in-use to [on].

**Comment by bwp91** [comment]
> as boring as could be

boring is normally good when it comes to things like this!

Just released v2.2.0 so the update should show in HOOBS within the next few hours or days or whatever. And hopefully no more errors for you both (or anyone else).

@fentonmatt and @sanist1 thanks for your patience on this and bringing it to my attention!

**Comment by fredthedorfman** [comment]
I updated to this new 2.2.0 version, doesn't crash homebridge but still get the errors and have to restart homebridge again anyway since the outlet switches become unresponsive. Going back to ver 1.6.2

**Comment by bwp91** [comment]
Hi @fredthedorfman 
Which errors did you receive in 2.2.0?
The plugin will routinely attempt to reconnect to any device that emits an error/becomes unresponsive.
The logs are a great way to see exactly what’s going on.
It would be nice for me to try sorting this out, but if 1.6.2 works better for you then of course using that version is a no-brainer :)

**Comment by fredthedorfman** [comment]
@bwp91 
Here's what I got on my homebridge logs on 2.2.0, first block of the log is the homekit automation trying to turn the outlet switch on

11/10/2020, 4:56:32 PM] [Belkin WeMo] [Wemo Mini] setting state and outlet-in-use to [on] error:
This callback function has already been called by someone else; it can only be called one time.
[11/10/2020, 4:56:32 PM] [Belkin WeMo] [Wemo Mini] Reverting HomeKit status due to error.
[11/10/2020, 4:56:32 PM] [Belkin WeMo] [Wemo Mini] setting state and outlet-in-use to [on] error:
Device uuid:Socket-1_0-22C8B1K010A06C has reported error ECONNREFUSED
[11/10/2020, 4:56:32 PM] [Belkin WeMo] [Wemo Mini] Reverting HomeKit status due to error.

Then I tried to go into homekit and manually turn it on and it gave the same errors below.  I had to restart homebridge to get it working again

[11/10/2020, 5:21:53 PM] [Belkin WeMo] [Wemo Mini] setting state and outlet-in-use to [on] error:
This callback function has already been called by someone else; it can only be called one time.
[11/10/2020, 5:21:53 PM] [Belkin WeMo] [Wemo Mini] Reverting HomeKit status due to error.
[11/10/2020, 5:21:53 PM] [Belkin WeMo] [Wemo Mini] setting state and outlet-in-use to [on] error:
connect ECONNREFUSED 192.168.0.185:49153
[11/10/2020, 5:21:53 PM] [Belkin WeMo] [Wemo Mini] Reverting HomeKit status due to error.
[11/10/2020, 5:21:57 PM] [Belkin WeMo] [Wemo Mini] setting state and outlet-in-use to [on] error:
This callback function has already been called by someone else; it can only be called one time.
[11/10/2020, 5:21:57 PM] [Belkin WeMo] [Wemo Mini] Reverting HomeKit status due to error.
[11/10/2020, 5:21:57 PM] [Belkin WeMo] [Wemo Mini] setting state and outlet-in-use to [on] error:
Device uuid:Socket-1_0-22C8B1K010A06C has reported error ECONNREFUSED
[11/10/2020, 5:21:57 PM] [Belkin WeMo] [Wemo Mini] Reverting HomeKit status due to error.

**Comment by bwp91** [comment]
@fredthedorfman out of interest which version of iOS/tvOS is your home hub device running?

**Comment by fredthedorfman** [comment]
@bwp91 14.2 for both

**Comment by bwp91** [comment]
A couple of other things I can suggest:

* removing the troublesome device from the homebridge cache so it is readded upon restart
* deleting the automation via the home app and re-creating it

**Comment by fredthedorfman** [comment]
@bwp91 
Alright I removed all 4 the cached accessories I have, deleted the automations and deleted the plugin completely.  I’m installing the 2.2.0 plugin back and configuring the automations again.  Automations will run tonight and tomorrow.

**Comment by bwp91** [comment]
@fredthedorfman 
I have *finally* be able to recreate the issue. So don't be disheartened if it still doesn't work.
Working on a fix!

**Comment by bwp91** [comment]
Okay, should be fixed in beta v2.3.0-4 if anyone is happy to try it out?

**Comment by fredthedorfman** [comment]
Got the latest beta installed, will test and let you know.

**Comment by sanist1** [comment]
Hi everything checks out ok for me on 2.3.0-5. no issues with automation

**Comment by fredthedorfman** [comment]
Went back to 1.6.2.  Still getting homebridge restart issues.  It could be an incompatibility with another plugin or something else but I'm just going to stick with what works.  Thanks for trying to help.

**Comment by bwp91** [comment]
@fredthedorfman first of all thanks for your patience with the betas and bearing with me.
i’m not totally happy that you’re still having the errors, as it’s highly unlikely it’s an incompatibility with another plugin. most likely something i have overlooked with the changes i made to v2.0.
Rest assured i will continue to look into this.

**Comment by bwp91** [comment]
Just to discount any other possibilities it would be interesting to hear if the errors are indeed resolved with you rolling back to v1.6

**Comment by fredthedorfman** [comment]
@bwp91 Good news, seems like everything is working for me.  I had more time to troubleshoot this over the weekend.  I ended up doing a  wipe / clean install of homebridge and node.js.  Only thing I saved was the config.json.  Was running 2.30.7 most of the day yesterday until this morning when I updated to your latest release of 2.30.  So far so good.  Automations running without problem and the switches in the Home app respond all the time now without homebridge getting stuck not responding.
